hconfig/mapl3hconfig_get_private.F90

```diff
@@ -0,0 +1,107 @@
+#include "MAPL_ErrLog.h"
+module mapl3hconfig_get_private
+   use :: esmf, only: ESMF_HConfig, ESMF_HConfigIsDefined, MAXSTRLEN => ESMF_MAXSTR
+   use :: esmf, only: ESMF_HConfigAsI4, ESMF_KIND_I4, ESMF_HConfigAsI8, ESMF_KIND_I8
+   use :: esmf, only: ESMF_HConfigAsR4, ESMF_KIND_R4, ESMF_HConfigAsR8, ESMF_KIND_R8
+   use :: esmf, only: ESMF_HConfigAsLogical, ESMF_HConfigAsString
+   use mapl_ErrorHandling
+   use mapl_KeywordEnforcer
+
+   implicit none
+
+   public :: MAXSTRLEN
+   public :: get_value
+
+   interface get_value
+      module procedure :: get_value_scalar
+   end interface get_value
+
+contains
+
+   subroutine get_value_scalar(hconfig, value, found, message, keystring, rc)
+      type(ESMF_HConfig), intent(inout) :: hconfig
+      class(*), intent(inout) :: value
+      logical, intent(out) :: found
+      character(len=:), allocatable, intent(inout) :: message
+      character(len=*), intent(in) :: keystring
+      integer, intent(out) :: rc
+
+      character(len=:), allocatable :: typestring
+      character(len=:), allocatable :: valuestring
+
+      integer :: status
+      integer :: ios
+      character(len=MAXSTRLEN) :: rawstring
+
+      found = ESMF_HConfigIsDefined(hconfig, keystring=keystring, _RC)
+      if(.not. found) then
+         _RETURN(_SUCCESS)
+      end if
```

Comment:
This is SUCCESS?

Reply:
Yes. It is, in essence, checking a boolean condition, and if true, returning a value.   **Checking** a boolean condition has three outcomes: true, false, error ("bottom"). I wanted to distinguish between: (1) the value is false, and (2) there was an error checking the condition. Our standard exception produces strings, stacktraces, and integers. So, determining the error requires parsing the message string or checking the integer value of the return code. Parsing the message to drive program logic is rather dangerous, and beyond 0 is success, we don't have code that looks up error conditions based on the integer return code. (Let me know if there is such code.)

If `default` (currently in the calling procedure) were mandatory, it would be impossible to not set the value, unless there were some other error condition. In that case, the flag that indicates if the value returned equals the default, would be mandatory. That flag is optional right now, and it is used for logging since we want to indicate if the value equals the default. I wanted to push that logic higher.

At the higher level, there is an `_FAIL` condition if: (1) the value is not found, (2) there is no default value, **and** (3) there is no `found` flag. (It's an `_ASSERT` statement.)

I don't like a mandatory default value because that invites passing in a default that may not be useful, just to pass in a default. The optional arguments should make it possible to handle the condition that the keystring is not found.

We could be stricter and say that the default is mandatory, and that the user needs to be careful to check that the value returned is a sensible value. Because the HConfig values are run-time, it's hard to safeguard against a user failing to include a keystring a file, especially since some users don't want a runtime error in that case.